text/0000-const-trait-impls.md

```diff
@@ -0,0 +1,1087 @@
+
+In contrast to other keywords like `unsafe` or `async` (that give you raw pointer derefs or `await` calls respectively),
+the `const` keyword on functions or blocks restricts what you can do within those functions or blocks.
+Thus the compiler historically used `host` as the internal inverse representation of `const` and `(const)` bounds.
+
+We generate a `ClauseKind::HostEffect` for every `const` or `(const)` bound.
+To mirror how some effectful languages represent such effects,
+I'm going to use `<Type as Trait>::k#constness` to allow setting whether the `constness` effect is "const" (disabled) or "conditionally" (generic).
+This is not comparable with other associated bounds like type bounds or const bounds, as the values the associated constness effect can
+take do neither have a usual hierarchy of trait bounds nor subtyping nor a concrete single value we can compare due to the following handling of those bounds:
+
+* There is no "disabled", as that is just the lack of a constness effect, meaning no `<Type as Trait>::k#constness` bound at all.
+* In contrast to other effect systems, we do not track the effect as a true generic parameter in the type system,
+  but instead explicitly convert all requirements of `Conditionally` bounds in always-const environments to `Const`.
+  * in other words: calling a `const fn<T: (const) Trait>()` in a const item or const block requires proving that the type used for `T` is `const`, as `(const)` can't refer to any conditionally const bound like it can within other const fns.
+
+While this could be modelled with generic parameters in the type system, that:
+
+* Has been attempted and is really complex (fragile) on the impl side and on the reasoning about things side.
+* Appears to permit more behaviours than are desirable (multiple such parameters, math on these parameters, ...), so they need to be prevented, adding more checks.
+* Is not necessary unless we'd allow much more complex kinds of bounds. So it can be kept open as a future possibility, but for now there's no need.
+* Does not quite work in Rust due to the constness then being early bound instead of late bound, cause all kinds of problems around closures and function calls.
+* Technically cause two entirely separate MIR bodies to be generated, one for where the effect is on and one where it is off. On top of that it then theoretically allows you to call the const MIR body from non-const code.
+
+Thus that approach was abandoned after proponents and opponents cooperated in trying to make the generic parameter approach work, resulting in all proponents becoming opponents, too.
+
+### Sites where `const Trait` bounds can be used
+
+Everywhere where non-const trait bounds can be written, but only for traits that have `(const)` methods.
+
+### Sites where `(const) Trait` bounds can be used
+
+* `const fn`
+* `(const) fn`
+* `impl const Trait` blocks
+* NOT in inherent impls, the individual `const fn` need to be annotated instead
```

Comment:
(Restating a point I tried to communicate in the now-resolved https://github.com/rust-lang/rfcs/pull/3762#discussion_r1990439966 thread)

This breaks a property that currently holds in Rust:

> If you delete `Trait for` from a trait `impl`, you are left with a perfectly valid inherent impl (disregarding coherence and default items not overridden).

Of course, we could always fix this backward-compatibly later, but we might as well have it work properly now.

Reply:
Also modulo visibility

Reply:
Without adding `pub` annotations, you would have an entirely private inherent impl, but that would still be completely valid.

Reply:
Also, as mentioned in that thread, modulo that the meaning of bounds in the impl can change.

Reply:
And also, directly to the point here, this doesn't work at all if you use associated types in the impl.

What we have here is something that maps to associated effects, so it's no surprise it's the same.  Sure, with associated types, you could manually normalize those out.  Here, we're not yet providing the desugaring syntax you'd need to do that.  But we later could, and then it would be no different in this respect.

Or, more speculatively, maybe we'd later allow `(const)` in inherent impls.  There, it'd map to an inherent associated effect (by analogy to inherent associated types).  There's not much point in that today, but maybe someday, with enough of the desugaring stabilized, there could be some use in this.